File: 0014-longest-common-prefix/0014-longest-common-prefix.js
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function(strs) {
    return recursiveLCP(strs, 0, strs.length - 1);
};

var recursiveLCP = function(strs, startIndex, endIndex) {
    if (startIndex === endIndex) return strs[startIndex];
    var mid = Math.floor((startIndex+endIndex)/2);
    var leftLCP = recursiveLCP(strs, startIndex, mid);
    var rightLCP = recursiveLCP(strs, mid+1, endIndex);
    return lcpComparison(leftLCP, rightLCP);
}

var lcpComparison = function(str1, str2) {
    var n = Math.min(str1.length, str2.length);
    var lcp = '';
    for (let i = 0; i < n; i++) {
        if (str1.charAt(i) === str2.charAt(i)) {
            lcp += str1.charAt(i);
        } else {
            return lcp;
        }
    }
    return lcp;
}

File: 0014-longest-common-prefix/0014-longest-common-prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        lcp = ''
        for i in range(min([len(s) for s in strs])):
            ogChar = strs[0][i]
            charMatches = True
            for s in strs:
                if s[i] == ogChar:
                    charMatches = charMatches and True
                else:
                    print('non matching char scenario')
                    return lcp
            if charMatches:
                lcp = lcp + ogChar
        print('the shortest string was the lcp')
        return lcp